File: bin/benchmarks/scanner.rb
#!/usr/bin/env ruby
#
# Copyright 2013-present Greg Hurrell. All rights reserved.
# Licensed under the terms of the BSD 2-clause license.

require_relative 'base.rb'

class CommandT::ProgressReporter
  def update t
    t + 1e9
  end
end

class MatcherBenchmark < CommandT::Benchmark
  TIMES = 5

  TYPE  = ENV.fetch 'TYPE', 'ruby'
  DIR   = ENV.fetch 'DIR', '/usr'

  def exec b
    scanner = CommandT::Scanner::FileScanner
      .for_scanner_type(TYPE)
      .new DIR,
        max_files: 1e9,
        git_scan_submodules: ENV.fetch('SUBMODULES', false),
        git_include_untracked: ENV.fetch('UNTRACKED', false)

    begin
      puts scanner.send :command
    rescue
    end

    b.report do
      scanner.flush
      scanner.c_paths
    end
  end
end

MatcherBenchmark.new.run

File: bin/benchmarks/scanner.lua
#!/usr/bin/env luajit

-- SPDX-FileCopyrightText: Copyright 2014-present Greg Hurrell and contributors.
-- SPDX-License-Identifier: BSD-2-Clause

local ffi = require('ffi')

local pwd = os.getenv('PWD')
local lua_directory = pwd .. '/' .. debug.getinfo(1).source:match('@?(.*/)') .. '../../lua'

package.path = lua_directory .. '/?.lua;' .. package.path
package.path = lua_directory .. '/?/init.lua;' .. package.path

local benchmark = require('wincent.commandt.private.benchmark')

benchmark({
  config = 'wincent.commandt.benchmark.configs.scanner',

  log = 'wincent.commandt.benchmark.logs.scanner',

  setup = function(config)
    local scanner = require(config.source)
    if scanner.name == 'watchman' then
      -- TODO: don't hardcode this, obviously...
      scanner.set_sockname('/opt/homebrew/var/run/watchman/wincent-state/sock')
    end
    return scanner
  end,

  skip = function(config)
    return config.skip_in_ci and os.getenv('CI')
  end,

  run = function(config, setup)
    local scanner = setup.scanner(pwd) -- For now, only Watchman wants pwd.
    for i = 1, scanner.count do
      ffi.string(scanner.candidates[i - 1].contents)
    end
  end,
})
